Sally7/RequestExecutor/ConcurrentRequestExecutor.cs

```diff
@@ -121,16 +129,35 @@ public async ValueTask<Memory<byte>> PerformRequest(ReadOnlyMemory<byte> request
 
                 using (IMemoryOwner<byte> mo = memoryPool.Rent(bufferSize))
                 {
-                    _ = await receiveSignal.WaitAsync(cancellationToken).ConfigureAwait(false);
-                    try
-                    {
-                        length = await reader.ReadAsync(mo.Memory, cancellationToken).ConfigureAwait(false);
-                    }
-                    finally
+                    // If we bail while reading we break the send-one-receive-one flow, so we might as well close right away.
+                    // There is minimal risk of closing connections while data was actually received but handling here
+                    // avoids registering on the cancellationToken on every socket call.
+#if NET5_0_OR_GREATER || NETSTANDARD2_1_OR_GREATER
+                    await
+#endif
+                    using (cancellationToken.MaybeUnsafeRegister(socket.Close))
                     {
-                        if (!receiveSignal.TryRelease())
+                        _ = await receiveSignal.WaitAsync(cancellationToken).ConfigureAwait(false);
+                        try
+                        {
+                            try
+                            {
+                                length = await reader.ReadAsync(mo.Memory, cancellationToken).ConfigureAwait(false);
+                            }
+                            catch (Exception) when (cancellationToken.IsCancellationRequested)
+                            {
+                                cancellationToken.ThrowIfCancellationRequested();
+
+                                // Should never happen, but required to satisfy compiler analysis of length initialization.
```

Comment:
I'd initialized the `length` in L128 with `0`, so that this statement can be removed.
* feels more natural (for me) than having an artificial exception
* reduced code size
* makes job for JIT easier

Reply:
I was actually quite happy with my solution, but I think yours still is better 😄

Reply:
Changed this per recommendation. I did figure another option, I could add `cancellationToken.Throw` as extension, add `[DoesNotReturn]` there and _maybe_ add the fallback exception as well (because someone could invoke that extension even when cancellation was not requested). Let me know what you think about that.

Reply:
Hm, to me this sounds like a bit of ceremony for an easy solution that you have now (with initialized length). I wouldn't do this via extension method.

Side note: AFAIK `[DoesNotReturn]` is used only by C# for nullability static analysis (so it has no real effect in this case).

Reply:
I believe `[DoesNotReturn]` actually does more, see for instance S7Connection.cs `GetExecutorOrThrow()` method.